### PR TITLE
Add Elasticsearch Log Exporter with tests

### PR DESCRIPTION
**Pull request by MarkSeufert**
This PR adds an Elasticsearch exporter for logs, which is a requirement from [issue #337](https://github.com/open-telemetry/opentelemetry-cpp/issues/337). The main features include:
- Specifying the host, port, and index/data stream of the Elasticsearch instance
- Sending an HTTP POST request to Elasticsearch using the [HTTP Client Interface](https://github.com/open-telemetry/opentelemetry-cpp/pull/385), and parsing the response message to ensure it was successfully received
- Sending batches of LogRecords using the [Bulk API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html)
- Timeout logic, such that the Export() method doesn't block indefinitely
- Unit tests for the above functionality

Before reviewing this PR, it should be noted that:
* This exporter current does not support CMake build as there is currently no CMake support for the Nlohmann/JSON library - will try to find a way around this using a backport for it (cc @maxgolov)
* Changes to logger.h and log_record.h can be ignore, as this PR will soon be changed to use a recordable instead of an API LogRecord once the [recordable PR](https://github.com/open-telemetry/opentelemetry-cpp/pull/438) gets merged. 

cc - @xukaren @alolita @lalitb @jsuereth @ThomsonTan

### CODE REVIEW

**Comment by lalitb** [comment]
Though I don't have strong opinion, would be good to get other's opinion for the right repo to merge this exporter PR, and other similar PRs - `main` or `contrib`. As per spec guidelines 
 - https://github.com/open-telemetry/opentelemetry-specification/blob/0229140bb13e8473f7c51513afd12e2a60618488/specification/library-guidelines.md - only limited set of exporters can go in main repo:
```
The SDK implementation should include the following exporters:

Jaeger.
Zipkin.
Prometheus.
OpenTelemetry Protocol (when the protocol is specified and approved).
Standard output (or logging) to use for debugging and testing as well as an input for the various log proxy tools.
In-memory (mock) exporter that accumulates telemetry data in the local memory and allows to inspect it (useful for e.g. unit tests).

Note: some of these support multiple protocols (e.g. gRPC, Thrift, etc). The exact list of protocols to implement in the exporters is TBD.

Other vendor-specific exporters (exporters that implement vendor protocols) should not be included in language libraries and should be placed elsewhere (the exact approach for storing and maintaining vendor-specific exporters will be defined in the future).
```

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (MarkSeufert) [reply exporters/elasticsearch/src/es_log_exporter.cc:153]: Sure thing, I'll move them there after I add the session manager as a private member.

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (MarkSeufert) [reply exporters/elasticsearch/src/es_log_exporter.cc:72]: Oh cool, I wasn't sure what this callback was used for but that sounds very useful. I'll add some functionality for this!

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (MarkSeufert) [reply exporters/elasticsearch/src/es_log_exporter.cc:147]: Yea, the only time we wouldn't want the `SendRequest()` method to be synchronous is if the main thread didn't care about the status of the response message. All the exporters have this blocking for response status requirement so unless the http client is going to be used outside of the exporters, changing the API would simplify this a bunch :)

File: exporters/CMakeLists.txt
Comment (ThomsonTan) [inline exporters/CMakeLists.txt:7]: Off by default and add an flag like `WITH_ELASTICSEARCH` to guard it?

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (lalitb) [inline exporters/elasticsearch/src/es_log_exporter.cc:248]: Is there any document defining the mapping between opentelementry log fields to elasic ?

**Comment by lalitb** [comment]
Also, please create a github issue for having an example on how to run this against elastic server. As we have for otlp:
https://github.com/open-telemetry/opentelemetry-cpp/tree/master/examples/otlp

File: exporters/elasticsearch/BUILD
Comment (jsuereth) [inline exporters/elasticsearch/BUILD:33]: Since you're baking the 'curl' client into the exporter itself, you need the copts to be part of the library above.

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (MarkSeufert) [reply exporters/elasticsearch/src/es_log_exporter.cc:248]: No I haven't seen any specs for this, but the Elasticsearch data schema is very flexible so I'm so I was keeping the data as close to the Log Data Model as possible. There's a recent issue for the collector about [supporting logs into their Metrics/Traces ES exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1800), so I'll ping them about it there.

File: exporters/elasticsearch/BUILD
Comment (MarkSeufert) [reply exporters/elasticsearch/BUILD:21]: Worked like a charm, thanks Josh :)

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:86]: nit: this map is already included in https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/logs/severity.h#L65

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:98]: nit: not sure if this would be better stored as a string? i.e. `std::string(name)`? (Same for `body` field)

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:114]: (same here, and in line 125, maybe a string for instead of `key.data()`?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h:109]: name variable `is_shutdown_` to follow Google naming convention?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h:105]: default timeout `std::chrono::microseconds::max()`?

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (xukaren) [inline exporters/elasticsearch/src/es_log_exporter.cc:65]: nit: add `{}` around if block?

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (xukaren) [inline exporters/elasticsearch/src/es_log_exporter.cc:54]: nit: is this mutex useful if it is only defined in scope of this function?

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (xukaren) [inline exporters/elasticsearch/src/es_log_exporter.cc:251]: nit: I think when data gets to elasticsearch, ideally the types should be preserved? ie. traceid/spanid/traceflags should be preserved as a hex value, not a string?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h:126]: same here, not sure if every single type should be converted to a string, or if original data types should be preserved when converted to json

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (MarkSeufert) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:86]: cool I didn't realize the recordable PR had this, definitely makes it simpler :)

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (xukaren) [inline exporters/elasticsearch/src/es_log_exporter.cc:66]: nit: add {} around if block?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h:114]: name variable `is_shutdown_` to follow Google naming convention?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h:110]: default timeout `std::chrono::microseconds::max()`?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:98]: nit: not sure if this would be better stored as a string? i.e. std::string(name)? (Same for body field)

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (xukaren) [inline exporters/elasticsearch/src/es_log_exporter.cc:55]: nit: is this mutex useful if it is only defined in scope of this function?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (MarkSeufert) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:98]: the data() method returns a `char *`, which is what a string is so I don't think it matters.

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (MarkSeufert) [reply exporters/elasticsearch/src/es_log_exporter.cc:55]: hmm it's needed for the condition variable `wait_for` method to work, but looks like traces does it [more condensed](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/src/trace/batch_span_processor.cc#L92) so good catch, I'll do it that way too

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (MarkSeufert) [reply exporters/elasticsearch/src/es_log_exporter.cc:66]: sure

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h
Comment (MarkSeufert) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h]: yup agreed, I'll take a look at switching it over to use the correct data types

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:36]: nit: maybe rename `JSONLogRecord`? Could be reused for a few other exporters as well

Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:36]: nit: maybe rename `JSONLogRecord`? Could maybe be used for other exporters?

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (MarkSeufert) [reply exporters/elasticsearch/src/es_log_exporter.cc]: 100% Agree about the types being preserved. For trace/span/flags though, I think it's more human readable if it is left as a string representation of the hex values, then if it was a raw binary blob.

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:91]: nit: I think `severity_num` and `severity_text` should both separated into 2 different fields and both added to the json object - following the spec's rec, having both is ideal :)

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (lalitb) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:36]: Don't think we can use it for others, as json format here would be specific to Elastic. Current name looks good in that case.

Comment (lalitb) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:36]: ok pls ignore my comment, I think there is no format defined for elastic search, and we are kind of using somewhat generic format here. Still I feel we can keep current name : )

Comment (xukaren) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:36]: haha yeah, i havent got any strong opinions on it - I agree it's a pretty generic format. so I'm wondering whether there's more use cases for it - maybe logging structured JSON data to a file would be useful? so those files could be used to be imported to databases or something? Im good with the name either way :)

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (MarkSeufert) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:91]: hmmm it might be overkill for the user, since the severity probably isn't the main focus of the log statement. Having one field that says the severity is probably fine, but if anyone else feels strongly I can change it :)

Comment (xukaren) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h]: wouldn't elasticsearch need to filter/query by severity? maybe the severity number would be easier to compare than the text then?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h
Comment (ThomsonTan) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_exporter.h:30]: `namespace json = nolhmann::json`?

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (lalitb) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:139]: Nit: It seems most of the code is repeated in these two functions `SetAttributes` and `SetResources`. Can we do instead with single function, either through some template function, or pass another argument depicting "attributes" or "resource".

**Review by lalitb — APPROVED** [review]
LGTM: If we can invest some time for example code as done for OTLP exporter, that would be great:
https://github.com/open-telemetry/opentelemetry-cpp/tree/master/examples/otlp

Or else please create a github issue for same.

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (MarkSeufert) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:139]: Yup good idea, I'll move it over into a function that takes in a string as the parameter.

**Comment by MarkSeufert** [comment]
I have an Elasticsearch example written in my team repo right now, but I'm holding off for PRing it until this one gets merged. I think Its the same situation for the ostream exporter too, so 2 more future PRs!

File: exporters/elasticsearch/BUILD
Comment (jsuereth) [inline exporters/elasticsearch/BUILD:21]: note to self:  Really want to clean this up... glad it's working though!

**Review by jsuereth — COMMENTED** [review]
AFAIk you only have two concurrency things happening here:

1. The HTTP libraries connection handling thread(s).
2. Your exporter, waiting to fire off a batch of data.

The threading code looks a little suspect (especially instantiating a local mutex).   Effectively you'll want some kind of memory barrier around access the shared variables between threads *or* you can use atomic variables to control that.

File: exporters/elasticsearch/src/es_log_exporter.cc
Comment (lalitb) [reply exporters/elasticsearch/src/es_log_exporter.cc:60]: jfyi - this code would become much simplified once we have synchronous request PR merged, as we won't need to wait for response anymore  : https://github.com/open-telemetry/opentelemetry-cpp/pull/448

This is how it would look like ( ref- zipkin exporter ) : https://github.com/open-telemetry/opentelemetry-cpp/blob/1039f4c2b947f3b3fd0880457ec77743fd7a80e7/exporters/zipkin/src/zipkin_exporter.cc#L62

Comment (MarkSeufert) [reply exporters/elasticsearch/src/es_log_exporter.cc:60]: Alright alright, I'll add a lock_guard around body_ and response_received_ so they can't be read and written at the same time. Yea, this is probably going to be replaced by Lalit's synchronous request, but for the sake of getting this merged ASAP we can always change it in a later issue/PR.

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (xukaren) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:53]: nit: does this work? or would `true` and `false` need to be wrapped with double quotes?

File: exporters/elasticsearch/test/es_log_exporter_test.cc
Comment (xukaren) [inline exporters/elasticsearch/test/es_log_exporter_test.cc:72]: Suggestion for this test: maybe cast `record` to an `ElasticSearchRecordable` object, so the `Get()` methods can be tested?

E.g:
```
auto es_record = unique_ptr<ElasticSearchRecordable >(static_cast<ElasticSearchRecordable *>(record.release()));

EXPECT_EQ(es_record["name"], "Timeout Log");
EXPECT_EQ(es_record["severity"], "Fatal");
etc.
```

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (MarkSeufert) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:53]: yea, instead of storing a string representation fo the bool it is storing an actual bool to preserve the datatype

File: exporters/elasticsearch/test/es_log_exporter_test.cc
Comment (MarkSeufert) [reply exporters/elasticsearch/test/es_log_exporter_test.cc:72]: yea good idea! I was assuming the nlohmann JSON library would work but better safe than sorry

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (xukaren) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:53]: oh cool cool, maybe `opentelemetry::nostd::get<bool>(value)` would work then? (nbd just wondering)

File: exporters/elasticsearch/test/es_log_exporter_test.cc
Comment (xukaren) [reply exporters/elasticsearch/test/es_log_exporter_test.cc:72]: yeah would be safer to test the `AttributeValue` types more thoroghly maybe?

**Review by jsuereth — APPROVED** [review]

**Pull request merged**
